File: frontend/src/services/pedidoService.js
import axios from "axios";

const PEDIDOS_URL = "http://localhost:8081/api/pedidos";

export const getAllPedidos = async () => {
  const { data } = await axios.get(PEDIDOS_URL);
  return data;
};

export const getPedidosByUserId = async (userId) => {
  const { data } = await axios.get(`${PEDIDOS_URL}/usuario/${userId}`);
  return data;
};

export const getPedidoById = async (pedidoId) => {
  try {
    const { data } = await axios.get(`${PEDIDOS_URL}/${pedidoId}`);
    return data;
  } catch (error) {
    console.error("Error al obtener detalles del pedido:", error);
    throw error;
  }
};

// Nuevas funcionalidades para actualización de estado
export const actualizarEstadoPedido = async (pedidoId, nuevoEstado) => {
  try {
    const { data } = await axios.put(`${PEDIDOS_URL}/${pedidoId}/estado`, {
      estado: nuevoEstado,
    });
    return data;
  } catch (error) {
    console.error("Error al actualizar estado del pedido:", error);
    throw error;
  }
};

export const actualizarEstadoPedidoForzado = async (
  pedidoId,
  nuevoEstado,
  password
) => {
  try {
    const { data } = await axios.put(
      `${PEDIDOS_URL}/${pedidoId}/estado/forzado`,
      {
        estado: nuevoEstado,
        password: password,
      }
    );
    return data;
  } catch (error) {
    console.error("Error al actualizar estado forzado del pedido:", error);
    throw error;
  }
};

export const getEstadosDisponibles = async () => {
  try {
    const { data } = await axios.get(`${PEDIDOS_URL}/estados`);
    return data;
  } catch (error) {
    console.error("Error al obtener estados disponibles:", error);
    throw error;
  }
};

export const getTransicionesDisponibles = async (estadoActual) => {
  try {
    const { data } = await axios.get(
      `${PEDIDOS_URL}/transiciones/${encodeURIComponent(estadoActual)}`
    );
    return data;
  } catch (error) {
    console.error("Error al obtener transiciones disponibles:", error);
    throw error;
  }
};

export const getAllOrders = getAllPedidos;

File: backend/src/main/java/com/comoencasa_backend/service/PedidoService.java
package com.comoencasa_backend.service;

import com.comoencasa_backend.dto.PedidoDTO;
import org.springframework.transaction.annotation.Transactional;


import java.util.List;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Optional;

public interface PedidoService {
    /**
     * Listar todos los pedidos (para el panel de administración)
     */
    List<PedidoDTO> findAll();

    /**
     * Obtener un pedido por su ID
     * 
     * @param pedidoId ID del pedido
     * @return Pedido encontrado
     * @throws IllegalArgumentException si el pedido no existe
     */
    PedidoDTO findById(Long pedidoId);

    /**
     * Obtener pedidos por ID de usuario
     * 
     * @param usuarioId ID del usuario
     * @return Lista de pedidos del usuario
     */
    List<PedidoDTO> obtenerPedidosPorUsuario(Long usuarioId);

    @Transactional
    PedidoDTO crearPedido(PedidoDTO pedidoDTO);

    /**
     * Actualizar estado de pedido con validaciones de flujo normal
     * 
     * @param pedidoId    ID del pedido
     * @param nuevoEstado Nuevo estado del pedido
     * @return Pedido actualizado
     * @throws IllegalArgumentException si la transición no es válida
     */
    @Transactional
    PedidoDTO actualizarEstadoPedido(Long pedidoId, String nuevoEstado);

    /**
     * Actualizar estado de pedido de forma forzada con confirmación especial
     * 
     * @param pedidoId    ID del pedido
     * @param nuevoEstado Nuevo estado del pedido
     * @param password    Contraseña de confirmación especial
     * @return Pedido actualizado
     * @throws IllegalArgumentException si la contraseña es incorrecta
     */
    @Transactional
    PedidoDTO actualizarEstadoPedidoForzado(Long pedidoId, String nuevoEstado, String password);

    /**
     * Validar si un estado es válido
     * 
     * @param estado Estado a validar
     * @return true si el estado es válido
     */
    boolean esEstadoValido(String estado);

    /**
     * Obtener transiciones disponibles desde un estado dado
     * 
     * @param estadoActual Estado actual del pedido
     * @return Lista de estados a los que se puede transicionar
     */
    List<String> getTransicionesDisponibles(String estadoActual);
    ByteArrayInputStream generarReporteVentasExcel(Optional<LocalDateTime> desde, Optional<LocalDateTime> hasta) throws IOException;

}